modules/transaction-payment/src/lib.rs

```diff
@@ -1202,11 +1200,12 @@ where
 			WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
 		};
 
-		Pallet::<T>::ensure_can_charge_fee(who, fee, reason);
+		let fee_surplus =
+			Pallet::<T>::ensure_can_charge_fee_with_call(who, fee, call).map_err(|_| InvalidTransaction::Payment)?;
 
-		// withdraw native currency as fee
-		match <T as Config>::Currency::withdraw(who, fee, reason, ExistenceRequirement::KeepAlive) {
-			Ok(imbalance) => Ok((fee, Some(imbalance))),
+		// withdraw native currency as fee, also consider surplus when swap from dex or pool.
+		match <T as Config>::Currency::withdraw(who, fee + fee_surplus, reason, ExistenceRequirement::KeepAlive) {
+			Ok(imbalance) => Ok((fee + fee_surplus, Some(imbalance), Some(fee_surplus))),
```

Comment:
return be returning `None` when `fee_surplus` is zero?

Reply:
no need to return `None` as it's only used in post_dispatch.